githubdelivery/src/main/java/io/dockstore/githubdelivery/GithubDeliveryS3Client.java

```diff
@@ -80,29 +101,121 @@ public static void main(String[] args) throws IOException {
         }
         return null;
     }
+    private InstallationRepositoriesPayload getGitHubInstallationRepositoriesPayloadByKey(String body, String key) throws IOException, NoSuchKeyException {
+        try {
+            InstallationRepositoriesPayload installationRepositoriesPayload;
+            installationRepositoriesPayload = MAPPER.readValue(body, InstallationRepositoriesPayload.class);
+            return installationRepositoriesPayload;
+        } catch (JsonSyntaxException e) {
+            exceptionMessage(e, String.format("Could not read github event from key %s", key), 1);
+        }
+        return null;
+    }
+    private void submitGitHubDeliveryEventsByDate(String date, WorkflowsApi workflowsApi) {
+
+        ListObjectsV2Request listObjectsV2Request = ListObjectsV2Request
+                .builder()
+                .bucket(bucketName)
+                .prefix(date)
+                .build();
+        ListObjectsV2Iterable listObjectsV2Iterable = s3Client.listObjectsV2Paginator(listObjectsV2Request);
+        for (ListObjectsV2Response response : listObjectsV2Iterable) {
+            response.contents().forEach((S3Object event) -> {
+                submitGitHubDeliveryEventsByKey(event.key(), workflowsApi);
+            });
+        }
+        LOG.info("Successfully submitted events for date {}", date);
+    }
+    private void submitGitHubDeliveryEventsByHour(String prefix, WorkflowsApi workflowsApi) {
+        ListObjectsV2Request listObjectsV2Request = ListObjectsV2Request
+                .builder()
+                .bucket(bucketName)
+                .prefix(prefix)
+                .build();
+        ListObjectsV2Iterable listObjectsV2Iterable = s3Client.listObjectsV2Paginator(listObjectsV2Request);
+        for (ListObjectsV2Response response : listObjectsV2Iterable) {
+            response.contents().forEach((S3Object event) -> {
+                submitGitHubDeliveryEventsByKey(event.key(), workflowsApi);
+            });
+        }
+        LOG.info("Successfully submitted events for date/hour {}", prefix);
+    }
+    private void submitGitHubDeliveryEventsByKey(String key, WorkflowsApi workflowsApi) {
+        String deliveryid = key.split("/")[2]; //since key is in YYYY-MM-DD/HH/deliveryid format
+        try {
+            String body = getObject(key);
+            JsonObject jsonObject = GSON.fromJson(body, JsonObject.class);
+            if (jsonObject.get("action").getAsString().equals("added") || jsonObject.get("action").getAsString().equals("removed")) {
+                InstallationRepositoriesPayload payload = getGitHubInstallationRepositoriesPayloadByKey(body, key);
+                if (payload != null) {
+                    workflowsApi.handleGitHubInstallation(payload, deliveryid);
+                } else {
+                    logReadError(key);
+                }
+
+            } else if (jsonObject.get("deleted").getAsBoolean()) {
```

Comment:
Will this work if `deleted` is not in the response (I don't know)? Or is `deleted` always in the response?

Reply:
I fixed the if case so it separates the installation events from the push events. In the push events, `deleted` is always in the response.